Documentation/proposals/202212-scrape-config.md

```diff
@@ -0,0 +1,106 @@
+  * [#3447](https://github.com/prometheus-operator/prometheus-operator/issues/3447)
+* Other docs:
+  * n/a
+
+This document aims at creating a lower level `ScrapeConfig` Custom Resource Definition that defines additional scrape configurations the Kubernetes way.
+
+# Why
+
+prometheus-operator misses a way to scrape external targets using CRD. Users have either been abusing the Probe CRD (#3447) or additionalScrapeConfig to do so.
+Furthermore, currently, there is a lot of code duplication due to the operator supporting several CRDs that generate scrape configurations. With the new `ScrapeConfig` CRD, it would be possible to consolidate some of that logic, where the other `*Monitor` CRDs could be migrated so that they create a ScrapeConfig resource that would ultimately be used by the operator to generate scrape configuration.
+
+## Pitfalls of the current solution
+
+Using `additionalScrapeConfig` comes with drawbacks:
+* Teams have to build an infrastructure to add scrape rules in a centralized manner, which creates a bottleneck since a single team becomes responsible for the configuration
+* There is no input validation, which can lead to an invalid prometheus configuration
+
+# Goals
+
+* Provide a way for users to self-service adding scrape targets
+* Consolidate the scrape configuration generation logic in a central point for other resources to use
+
+## Audience
+
+* Users who serve Prometheus as a service and want to have their customers autonomous in defining scrape configs
+* Users who want to manage scrape configs the same way as for services running within the Kubernetes cluster
+* Users who want a supported Kubernetes way of scraping targets outside the Kubernetes cluster
+
+# Non-Goals
+
+* This proposal doesn't aim at covering all the fields in [`<scrape_config>`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config). Specifically, no service discovery other than `static_configs` or `file_sd_configs` should be implemented at first.
+* refactoring of the other CRDs is not in scope for the first version
+
+# How
+
+As described by [@aulig in #2787](https://github.com/prometheus-operator/prometheus-operator/issues/2787#issuecomment-559776221), we will create a new `ScrapeConfig` CRD, this config will act the same as the other CRDs and append scrape configurations to the configuration. `ScrapeConfig` will allow for any scraping configuration, while the other CRDs provide sane defaults. This will allow for isolated testing of the new `ScrapeConfig` CRD.
+
+```mermaid
+graph TD;
+    ServiceMonitor-->prometheusConfig
+    PodMonitor-->prometheusConfig
+    ScrapeConfig-->prometheusConfig
+    prometheusConfig
+```
+
+Using a pseudo custom resource definition, we should have the following:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: ScrapeConfig
+metadata:
+  name: my-scrape-config
+  namespace: system-monitoring
+  labels:
+    test: value
+spec:
```

Comment:
I wonder if we shouldn't also include HTTP SD config from the start. It could be useful to experiment with target sharding techniques such as the OpenTelemetry collector's [target allocator](https://github.com/open-telemetry/opentelemetry-operator#target-allocator).

Reply:
We can, it shouldn't be too hard and is generic enough to help a lot of people from the start. I would limit to those (static/file/http) on the first step though and add other SDs as we go. This should be decoupled from the refactoring effort. Thoughts?

Reply:
I've added httpSDConfig to the list

Reply:
https://github.com/prometheus-operator/prometheus-operator/pull/5279/files#diff-a6321575a64799dcf97cfe7e62ce050189f50e6cd33fc6cc495f118f6cfefcddR143-R144